logger/storage/file/adapter.go

```diff
@@ -0,0 +1,130 @@
+package file
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+type adapter struct {
+	logRoot string
+	files   map[string]*os.File
+	mutex   sync.Mutex
+}
+
+// NewStorageAdapter returns a pointer to a new instance of a file-based storage.Adapter.
+func NewStorageAdapter(logRoot string) (*adapter, error) {
+	src, err := os.Stat(logRoot)
+	if err != nil {
+		return nil, fmt.Errorf("Directory %s does not exist", logRoot)
+	}
+	if !src.IsDir() {
+		return nil, fmt.Errorf("%s is not a directory", logRoot)
+	}
+	return &adapter{logRoot: logRoot, files: make(map[string]*os.File)}, nil
+}
+
+// Write adds a log message to to an app-specific log file
+func (a *adapter) Write(app string, message string) error {
+	// Check first if we might actually have to add to the map of file pointers so we can avoid
+	// waiting for / obtaining a lock unnecessarily
+	f, ok := a.files[app]
+	if !ok {
+		// Ensure only one goroutine at a time can be adding a file pointer to the map of file
+		// pointers
+		a.mutex.Lock()
+		defer a.mutex.Unlock()
+		f, ok = a.files[app]
+		if !ok {
+			var err error
+			f, err = a.getFile(app)
+			if err != nil {
+				return err
+			}
+			a.files[app] = f
+		}
+	}
+	if _, err := f.WriteString(message + "\n"); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Read retrieves a specified number of log lines from an app-specific log file
+func (a *adapter) Read(app string, lines int) ([]string, error) {
+	if lines <= 0 {
+		return []string{}, nil
+	}
+	filePath := a.getFilePath(app)
+	exists, err := fileExists(filePath)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, fmt.Errorf("Could not find logs for '%s'", app)
+	}
+	logBytes, err := exec.Command("tail", "-n", strconv.Itoa(lines), filePath).Output()
+	if err != nil {
+		return nil, err
+	}
+	logStrs := strings.Split(string(logBytes), "\n")
+	return logStrs[:len(logStrs)-1], nil
+}
+
+func (a *adapter) Destroy(app string) error {
+	filePath := a.getFilePath(app)
```

Comment:
You're right that I was mistakenly not removing the file pointer from `a.files`.  That was an oversight.  Originally, this was a bit more like the old logger and any given file was opened anew anytime a message was written.  (There was no map of file pointers then.)  That seemed absurd to me, so I added that optimization to open files and keep them open in a map, but overlooked the change that had to be made in `Destroy()` to clean that up.  Nice catch.  I'm fixing this.

Re: possible race conditions-- I suppose the possibility exists that a log message for a given application could arrive _after_ logs for that application have been destroyed.  It will just create a new log for that application is all.  I don't think there's much we can do about that.  Although probably not impossible, it's not a likely scenario because when an app is destroyed, the controller destroy's the application's containers _before_ it signals the logger to destroy that app's logs.